### PR TITLE
 Add HttpResponseStatusFuture to notify server connector error

### PR DESCRIPTION
**Pull request by chamil321**
## Automation tests
 - Unit tests 
   > Will be added in a different PR
 - Integration tests
   > Will be added in a different PR

### CODE REVIEW

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java [hunk 1]
```diff
@@ -39,11 +39,13 @@
     private ChannelHandlerContext sourceContext;
     private RequestDataHolder requestDataHolder;
     private HandlerExecutor handlerExecutor;
+    private HTTPCarbonMessage requestMsg;
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: shall we say inboundRequestMsg

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java [hunk 2]
```diff
@@ -65,6 +67,13 @@ public void onMessage(HTTPCarbonMessage httpResponseMessage) {
                     this.sourceContext.channel().eventLoop().execute(() -> {
                 if (Util.isLastHttpContent(httpContent)) {
                     ChannelFuture future = sourceContext.writeAndFlush(httpContent);
+                    HttpResponseStatusFuture responseStatusFuture = requestMsg.getHttpResponseStatusFuture();
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: shall we say getHttpOutboundRespStatusFuture

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java [hunk 3]
```diff
@@ -65,6 +67,13 @@ public void onMessage(HTTPCarbonMessage httpResponseMessage) {
                     this.sourceContext.channel().eventLoop().execute(() -> {
                 if (Util.isLastHttpContent(httpContent)) {
                     ChannelFuture future = sourceContext.writeAndFlush(httpContent);
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: shall we say OutboundChannelFuture

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java [hunk 4]
```diff
@@ -65,6 +67,13 @@ public void onMessage(HTTPCarbonMessage httpResponseMessage) {
...
+                    HttpResponseStatusFuture responseStatusFuture = requestMsg.getHttpResponseStatusFuture();
+                    future.addListener(genericFutureListener -> {
+                        if (genericFutureListener.cause() != null) {
+                            responseStatusFuture.notifyHttpListener(genericFutureListener.cause());
+                        }
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: Shouldn't it be inside the "else" part ?

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.concurrent.Semaphore;
+
+/**
+ * Implementation of the response status future.
+ */
+public class HttpResponseStatusFuture implements HttpResponseFuture {
+
+    private HTTPCarbonMessage httpCarbonMessage;
+    private Throwable throwable, status;
+    private volatile Semaphore executionWaitSem;
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java]: Any special reason for making it volatile ?

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    @Override
+    public void notifyHttpListener(Throwable throwable) {
+        this.throwable = throwable;
+        if (executionWaitSem != null) {
+            executionWaitSem.release();
+        }
+    }
+
+    public Throwable sync() throws InterruptedException {
+        executionWaitSem = new Semaphore(0);
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java:62]: Can I know the purpose of this logic ?

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java [hunk 7]
```diff
@@ -53,8 +54,9 @@
 
     private MessagingException messagingException = null;
     private MessageDataSource messageDataSource;
-    private ServerConnectorFuture serverConnectorFuture = new HttpWsServerConnectorFuture();
     private MessageFuture messageFuture;
+    private final ServerConnectorFuture serverConnectorFuture = new HttpWsServerConnectorFuture();
+    private final HttpResponseStatusFuture httpResponseStatusFuture = new HttpResponseStatusFuture();
```
Comment (shafreenAnfar) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java]: Shall we say httpOutboundRespFuture and httpOutboundRespStatusFuture

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java (hunk 4 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: ack

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java (hunk 2 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: ack

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java (hunk 3 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: ack

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java (hunk 1 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseListener.java]: ack

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java (hunk 5 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpResponseStatusFuture.java]: Removed as semaphore is visible between threads

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java (hunk 7 above)
Comment (chamil321) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/message/HTTPCarbonMessage.java]: ack

**Review by shafreenAnfar — APPROVED** [review]

**Pull request merged**